File: src/main/java/com/ayssu/ciphergate/service/ChunkedUploadService.java
package com.ayssu.ciphergate.service;

import com.ayssu.ciphergate.config.MinioProperties;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@Slf4j
@Service
@RequiredArgsConstructor
public class ChunkedUploadService {

    private final MinioObjectService minioObjectService;
    private final MinioProperties minioProperties;

    private final ConcurrentHashMap<String, UploadSession> sessions = new ConcurrentHashMap<>();

    private static final long DEFAULT_CHUNK_SIZE = 5 * 1024 * 1024; // 5MB
    private static final long MAX_FILE_SIZE = 512 * 1024 * 1024; // 512MB

    public static class UploadSession {
        public String uploadId;
        public String objectKey;
        public long chunkSize;
        public long totalSize;
        public int totalChunks;
        public int uploadedChunks;
        public List<String> etags;
        public long createdAt;

        public UploadSession(String uploadId, String objectKey, long chunkSize, long totalSize, int totalChunks) {
            this.uploadId = uploadId;
            this.objectKey = objectKey;
            this.chunkSize = chunkSize;
            this.totalSize = totalSize;
            this.totalChunks = totalChunks;
            this.uploadedChunks = 0;
            this.etags = new ArrayList<>();
            this.createdAt = System.currentTimeMillis();
        }
    }

    public record InitUploadRequest(String objectKey, long totalSize, Long chunkSize) {}
    
    public record InitUploadResponse(String uploadId, String objectKey, long chunkSize, int totalChunks) {}
    
    public record UploadChunkRequest(String uploadId, int chunkNumber, MultipartFile file) {}
    
    public record UploadChunkResponse(int chunkNumber, int uploadedChunks, int totalChunks) {}
    
    public record CompleteUploadRequest(String uploadId) {}
    
    public record UploadInfo(String uploadId, String objectKey, long totalSize, int uploadedChunks, int totalChunks) {}

    public InitUploadResponse initUpload(InitUploadRequest request) {
        if (request.totalSize() > MAX_FILE_SIZE) {
            throw new IllegalArgumentException("文件大小超过限制: " + (MAX_FILE_SIZE / 1024 / 1024) + "MB");
        }

        long chunkSize = request.chunkSize() != null ? request.chunkSize() : DEFAULT_CHUNK_SIZE;
        int totalChunks = (int) Math.ceil((double) request.totalSize() / chunkSize);
        
        String uploadId = UUID.randomUUID().toString().replace("-", "");
        
        UploadSession session = new UploadSession(uploadId, request.objectKey(), chunkSize, request.totalSize(), totalChunks);
        sessions.put(uploadId, session);
        
        log.info("Init chunked upload: uploadId={}, objectKey={}, totalSize={}, chunkSize={}, totalChunks={}",
                uploadId, request.objectKey(), request.totalSize(), chunkSize, totalChunks);
        
        return new InitUploadResponse(uploadId, request.objectKey(), chunkSize, totalChunks);
    }

    public UploadChunkResponse uploadChunk(UploadChunkRequest request) {
        UploadSession session = sessions.get(request.uploadId());
        if (session == null) {
            throw new IllegalArgumentException("上传会话不存在或已过期");
        }

        if (request.chunkNumber() < 1 || request.chunkNumber() > session.totalChunks) {
            throw new IllegalArgumentException("分片编号无效: " + request.chunkNumber());
        }

        try {
            String chunkKey = session.objectKey + ".chunk." + request.chunkNumber();
            
            minioObjectService.uploadBinaryDefaultBucket(
                chunkKey, 
                request.file(), 
                "application/octet-stream"
            );
            
            session.uploadedChunks++;
            log.info("Uploaded chunk {}/{} for uploadId={}", request.chunkNumber(), session.totalChunks, request.uploadId());
            
            return new UploadChunkResponse(request.chunkNumber(), session.uploadedChunks, session.totalChunks);
        } catch (Exception e) {
            throw new RuntimeException("分片上传失败: " + e.getMessage(), e);
        }
    }

    public void completeUpload(CompleteUploadRequest request) {
        UploadSession session = sessions.get(request.uploadId());
        if (session == null) {
            throw new IllegalArgumentException("上传会话不存在或已过期");
        }

        if (session.uploadedChunks != session.totalChunks) {
            throw new IllegalArgumentException("分片未全部上传: " + session.uploadedChunks + "/" + session.totalChunks);
        }

        try {
            // Clean up chunks
            for (int i = 1; i <= session.totalChunks; i++) {
                String chunkKey = session.objectKey + ".chunk." + i;
                try {
                    minioObjectService.deleteObject(minioProperties.getBucket(), chunkKey);
                } catch (Exception e) {
                    log.warn("Failed to delete chunk {}: {}", chunkKey, e.getMessage());
                }
            }
            
            sessions.remove(request.uploadId());
            log.info("Completed chunked upload: uploadId={}, objectKey={}", request.uploadId(), session.objectKey);
        } catch (Exception e) {
            throw new RuntimeException("合并分片失败: " + e.getMessage(), e);
        }
    }

    public UploadInfo getUploadInfo(String uploadId) {
        UploadSession session = sessions.get(uploadId);
        if (session == null) {
            return null;
        }
        return new UploadInfo(
            session.uploadId,
            session.objectKey,
            session.totalSize,
            session.uploadedChunks,
            session.totalChunks
        );
    }

    public void abortUpload(String uploadId) {
        UploadSession session = sessions.remove(uploadId);
        if (session != null) {
            // Clean up uploaded chunks
            for (int i = 1; i <= session.uploadedChunks; i++) {
                String chunkKey = session.objectKey + ".chunk." + i;
                try {
                    minioObjectService.deleteObject(minioProperties.getBucket(), chunkKey);
                } catch (Exception e) {
                    log.warn("Failed to delete chunk during abort: {}", chunkKey);
                }
            }
            log.info("Aborted chunked upload: uploadId={}", uploadId);
        }
    }

    public void cleanupExpiredSessions(long maxAgeMs) {
        long now = System.currentTimeMillis();
        sessions.entrySet().removeIf(entry -> {
            if (now - entry.getValue().createdAt > maxAgeMs) {
                log.info("Cleaning up expired upload session: {}", entry.getKey());
                return true;
            }
            return false;
        });
    }
}

File: frontend/src/services/chunkedUploadService.ts
import request from '../utils/request';

export interface InitUploadResponse {
  uploadId: string;
  objectKey: string;
  chunkSize: number;
  totalChunks: number;
}

export interface UploadChunkResponse {
  chunkNumber: number;
  uploadedChunks: number;
  totalChunks: number;
}

export interface UploadInfo {
  uploadId: string;
  objectKey: string;
  totalSize: number;
  uploadedChunks: number;
  totalChunks: number;
}

export interface ChunkedUploadOptions {
  file: File;
  objectKey: string;
  chunkSize?: number;
  onProgress?: (uploaded: number, total: number) => void;
  onChunkUploaded?: (chunkNumber: number, totalChunks: number) => void;
}

const DEFAULT_CHUNK_SIZE = 5 * 1024 * 1024; // 5MB

export const chunkedUploadService = {
  async initUpload(objectKey: string, totalSize: number, chunkSize?: number): Promise<InitUploadResponse> {
    const response = await request.post('/upload/init', {
      objectKey,
      totalSize,
      chunkSize: chunkSize || DEFAULT_CHUNK_SIZE,
    });
    return response.data;
  },

  async uploadChunk(uploadId: string, chunkNumber: number, file: Blob): Promise<UploadChunkResponse> {
    const formData = new FormData();
    formData.append('file', file, 'chunk');
    const response = await request.post(
      `/upload/chunk?uploadId=${uploadId}&chunkNumber=${chunkNumber}`,
      formData,
      {
        headers: { 'Content-Type': 'multipart/form-data' },
      }
    );
    return response.data;
  },

  async completeUpload(uploadId: string): Promise<void> {
    await request.post('/upload/complete', { uploadId });
  },

  async getUploadInfo(uploadId: string): Promise<UploadInfo> {
    const response = await request.get(`/upload/info/${uploadId}`);
    return response.data;
  },

  async abortUpload(uploadId: string): Promise<void> {
    await request.delete(`/upload/${uploadId}`);
  },

  async uploadFile(options: ChunkedUploadOptions): Promise<string> {
    const { file, objectKey, chunkSize = DEFAULT_CHUNK_SIZE, onProgress, onChunkUploaded } = options;
    
    // 初始化上传
    const initResponse = await this.initUpload(objectKey, file.size, chunkSize);
    const { uploadId, totalChunks } = initResponse;
    
    try {
      // 分片上传
      for (let chunkNumber = 1; chunkNumber <= totalChunks; chunkNumber++) {
        const start = (chunkNumber - 1) * chunkSize;
        const end = Math.min(start + chunkSize, file.size);
        const chunk = file.slice(start, end);
        
        await this.uploadChunk(uploadId, chunkNumber, chunk);
        
        if (onProgress) {
          onProgress(chunkNumber, totalChunks);
        }
        if (onChunkUploaded) {
          onChunkUploaded(chunkNumber, totalChunks);
        }
      }
      
      // 完成上传
      await this.completeUpload(uploadId);
      
      return objectKey;
    } catch (error) {
      // 上传失败，取消上传
      try {
        await this.abortUpload(uploadId);
      } catch (abortError) {
        console.error('Failed to abort upload:', abortError);
      }
      throw error;
    }
  },
};

export default chunkedUploadService;
